base_layer/core/tests/support/simple_block_chain.rs

```diff
@@ -0,0 +1,141 @@
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use chrono::Duration;
+use merklemountainrange::mmr::*;
+use rand::OsRng;
+use tari_core::{block::*, blockheader::*, pow::*, transaction::*, types::*};
+use tari_crypto::{keys::SecretKey, ristretto::*};
+use tari_utilities::hash::Hashable;
+
+/// This is used to represent a block chain in memory for testing purposes
+pub struct SimpleBlockChain {
+    blocks: Vec<Block>,
+    headers: MerkleMountainRange<BlockHeader, SignatureHash>,
+    utxos: MerkleMountainRange<TransactionInput, SignatureHash>,
+    kernels: MerkleMountainRange<TransactionKernel, SignatureHash>,
+}
+impl SimpleBlockChain {
+    /// This will create a new test block_chain
+    pub fn new(amount: u32) -> SimpleBlockChain {
+        let mut chain = SimpleBlockChain::default();
+
+        let mut rng = OsRng::new().unwrap();
+        // create gen block
+        let priv_key = PrivateKey::random(&mut rng);
+        let block = BlockBuilder::new().create_coinbase(priv_key).build_with_pow();
+        chain.processes_new_block(block);
+
+        // lets mine some more blocks
+        for _i in 1..(amount) {
+            let priv_key = PrivateKey::random(&mut rng);
+            let header = chain.generate_new_header();
+            let block = BlockBuilder::new()
+                .with_header(header)
+                .create_coinbase(priv_key)
+                .build_with_pow();
+            chain.processes_new_block(block);
+        }
+        chain
+    }
+
+    pub fn add(&mut self, amount: u32) {
+        let mut rng = OsRng::new().unwrap();
+        for _i in 0..(amount) {
+            let priv_key = PrivateKey::random(&mut rng);
```

Comment:
Might be worth keeping these private_keys in a separate structure or something so that you can spend the coinbase tx's while working with this blockchain

Reply:
Next PR will address that, already coded that in

Reply:
`/// This struct is used to keep track of what the value and private key of a UTXO is.
pub struct SpendInfo {
    pub key: PrivateKey,
    pub value: u64,
}`